kotsadm/pkg/handlers/airgap.go

```diff
@@ -90,48 +288,100 @@ func updateAppFromAirgap(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		logger.Error(err)
-		w.WriteHeader(500)
+		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
 	var registryHost, namespace, username, password string
 	registryHost, username, password, err = getKurlRegistryCreds()
 	if err != nil {
 		logger.Error(err)
-		w.WriteHeader(500)
+		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
 	// if found kurl registry creds, use kurl registry
 	if registryHost != "" {
 		namespace = pendingApp.Slug
 	} else {
-		registryHost = r.FormValue("registryHost")
-		namespace = r.FormValue("namespace")
-		username = r.FormValue("username")
-		password = r.FormValue("password")
+		registryHost = createAppFromAirgapRequest.RegistryHost
+		namespace = createAppFromAirgapRequest.Namespace
+		username = createAppFromAirgapRequest.Username
+		password = createAppFromAirgapRequest.Password
 	}
 
-	airgapBundle, _, err := r.FormFile("file")
+	identifier := mux.Vars(r)["identifier"]
+	airgapBundlePath := getAirgapBundlePath(identifier)
+
+	totalChunksStr := mux.Vars(r)["totalChunks"]
+	totalChunks, err := strconv.ParseInt(totalChunksStr, 10, 64)
 	if err != nil {
-		logger.Error(err)
-		w.WriteHeader(500)
+		logger.Error(errors.New("failed to parse total chunks number as integer"))
+		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
 
 	go func() {
-		defer airgapBundle.Close()
-		if err := airgap.CreateAppFromAirgap(pendingApp, airgapBundle, registryHost, namespace, username, password); err != nil {
-			logger.Error(err)
+		if err := airgap.CreateAppFromAirgap(pendingApp, airgapBundlePath, registryHost, namespace, username, password); err != nil {
+			logger.Error(errors.Wrap(err, "failed to create app from airgap bundle"))
+			return
+		}
+		// app created successfully, we can remove the airgap bundle
+		if err := cleanUp(identifier, totalChunks); err != nil {
+			logger.Error(errors.Wrap(err, "failed to clean up"))
 		}
 	}()
 
 	createAppFromAirgapResponse := CreateAppFromAirgapResponse{}
 
-	JSON(w, 202, createAppFromAirgapResponse)
+	JSON(w, http.StatusOK, createAppFromAirgapResponse)
+}
+
+func getChunkKey(uploadedFileIdentifier string, chunkNumber int64) string {
+	return fmt.Sprintf("%s_part_%d", uploadedFileIdentifier, chunkNumber)
+}
+
+func getAirgapBundlePath(uploadedFileIdentifier string) string {
+	return filepath.Join(os.TempDir(), fmt.Sprintf("%s.%s", uploadedFileIdentifier, "airgap"))
+}
+
+func isUploadComplete(uploadedFileIdentifier string, totalChunks int64) bool {
+	isUploadComplete := true
+
+	var i int64
+	for i = 1; i <= totalChunks; i++ {
+		chunkKey := getChunkKey(uploadedFileIdentifier, i)
+		if _, ok := uploadedAirgapBundleChunks[chunkKey]; !ok {
+			isUploadComplete = false
+		}
+	}
+
+	return isUploadComplete
+}
+
+func cleanUp(uploadedFileIdentifier string, totalChunks int64) error {
+	var i int64
+	for i = 1; i <= totalChunks; i++ {
+		chunkKey := getChunkKey(uploadedFileIdentifier, i)
+		delete(uploadedAirgapBundleChunks, chunkKey)
```

Comment:
Is `uploadedFileIdentifier` unique across different uploads?  It might be better to clean up this map when upload starts, at the same time file is created.

Reply:
it's generated based on the file name and size. if it's clean on start, then there won't be resumable uploads, no? (in case the upload was interrupted)

Reply:
it's always the same for the same file name and size

Reply:
it's always the same for the same file name and size